README.md
# 原创性检测服务器
|项目名称|项目描述|项目功能|项目特点|
|:----:|:----:|:----:|:----:|
|摄影作品原创性检测服务器|基于百度智云对摄影作品进行原创性检测|1.原创性检测</br> 2.日志管理</br> 3.监控与报警|原创性检测内容全面|
# 项目目录结构
```shell
├─assets // 静态资源
│  ├─images
│  ├─scripts
│  └─styles
├─controllers// 路由
│  ├─imageDetection.js     // 原创性检测路由
│  ├─log.js                // 日志管理路由
│  ├─logHelper.js          // 日志管理帮助文件
│  └─monitor.js            // 状态监控路由
├─docs// 文档
│  ├─History.md             // 项目修改历史文档
│  ├─INSTALL.md             // 项目安装文档
│  ├─常见Web攻击.md          // 常见网络攻击文档
│  ├─项目说明文档.md          // 项目说明文档
│  └─images                 // 文档中的图片 
├─logs// 日志
│  ├─imageDetection
│  ├─log
│  ├─monitor
│  └─process
├─node_modules// npm 模块包
├─test// 测试代码
│  ├─imageDetectionTest
│  |  ├─imageDetectionClient.js // 原创性检测服务测试
│  │  └─testImage
│  ├─logTest
│  │  └─logClient.js            // 日志管理服务测试
│  ├─moduleTest
│  │  └─diskInfo.js             // diskInfo模块测试
│  └─monitorTest
│     └─monitorClient.js        // 状态监控服务测试
├─app.js        // 项目入口
├─config.json   // 项目配置文件
├─nodemon.json  // nodemon配置文件
├─package.json  // npm 模块配置
├─README.md     // 说明文件--快速开始
└─views
```

# 快速开始
1. 安装所需要的依赖包
`npm install`
2. 配置相关参数
`config.json`
3. 调试模式启动项目
`npm dev`
4. 测试项目
`npm test`
# 了解详细内容请查看docs

docs/History.md
# 项目改动历史

|   日期    | 内容 | 人员 | 备注 |
| :-------: | :-------: | :-------: | :-------: |
|2021-9-23|项目初始化|cups|...|
|2021-9-24|https|cups|...|
|2021-9-27|功能代码编写|cups|...|
|2021-9-28|pm2部署测试|cups|...|



test/moduleTest/json.js

let data = {"name":"user1@cpoa.com","mspid":"CPOAMSP","roles":null,"affiliation":"","enrollmentSecret":"","enrollment":{"signingIdentity":"c5350ba05b401f93ce96679d272e546e525812e5ebc5b43c10dc0264527f8d7d","identity":{"certificate":"-----BEGIN CERTIFICATE-----\nMIICATCCAaegAwIBAgIRAMIjvX0uoniJEpbC7Z9NpmYwCgYIKoZIzj0EAwIwWjEL\nMAkGA1UEBhMCQ04xDjAMBgNVBAgTBUhlTmFuMRIwEAYDVQQHEwlaaGVuZ1pob3Ux\nETAPBgNVBAoTCGNwb2EuY29tMRQwEgYDVQQDEwtjYS5jcG9hLmNvbTAeFw0yMTA4\nMDUwOTQ5MDBaFw0zMTA4MDMwOTQ5MDBaMFsxCzAJBgNVBAYTAkNOMQ4wDAYDVQQI\nEwVIZU5hbjESMBAGA1UEBxMJWmhlbmdaaG91MQ8wDQYDVQQLEwZjbGllbnQxFzAV\nBgNVBAMMDlVzZXIxQGNwb2EuY29tMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE\n6r3Y9SygkZRNCWlwHflS+DTHyK8y9V3t/kqhz5OSOGQuWVXu1NTs85C6fxtoN2d4\nb64RyjebI8BgWTISUJbnmaNNMEswDgYDVR0PAQH/BAQDAgeAMAwGA1UdEwEB/wQC\nMAAwKwYDVR0jBCQwIoAgW8oSekHXEeVZWx8f3cPSmDmL6FyKQb8VAIgL73h+sIsw\nCgYIKoZIzj0EAwIDSAAwRQIhANldd4Kkxxh5Y8uBFFaQfwGZbrsl3cPYB77KiqEN\nQRq5AiARrai8jp3/ZvQYz9EzjOS25SQxaUEYP28WRjFYL165/w==\n-----END CERTIFICATE-----"}}}



console.log(data.enrollment.identity.certificate)



test/logTest/logClient.js
const https = require('https');
const querystring = require('querystring');
const config = require('../../config.json');

const sendRequest = (path='/', method='GET' , data = {}) => {
    return new Promise((resolve, reject) => {
        let message = '';
        if(data!== undefined){
            message = querystring.stringify(data);
        }
        // console.log(message);
        const options = {
            host: config.host,                      // ip
            port: config.port,                      // 端口
            path: path,
            method: method,
            headers:{
                'Content-Type':'application/x-www-form-urlencoded',
                'Content-Length':message.length
            },
            key: config.testPrivateKey,             // 私钥
            cert: config.testPublicKey,             // 公钥
            ca: config.caCertificateList[0],        // 为服务器颁发证书的CA列表
            agent: false,                           // 不使用代理
            rejectUnauthorized: false               // 双向认证
        };
        let req = https.request(options, (res) => {
            // console.log('client connected', res.connection.authorized ? 'authorized' : 'unauthorized');
            let statusCode = res.statusCode;
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                // console.log(statusCode, data);
                resolve({statusCode, data});
            });
        });
        req.on('error', (error) => {
            reject(error);
        });
        req.write(message);
        req.end();
    });
}


let conmmands = ['getAllFileDir', 'downloadOne', 'downloadAll', 'deleteOne', 'deleteAll'];
/**
 * 日志管理--getAllFileDir
 * **/

const main = async ()=>{
    let dir;
    await sendRequest(path='/log', method='POST' , data = {command:'getAllFileDir'}).then((response) => {

        if (response.statusCode === 200){
            console.log('statusCode:', response.statusCode);
            console.log('message:', JSON.parse(response.data).message);
            dir = JSON.parse(response.data).message;
        }else {
            console.log('statusCode:', response.statusCode);
            console.log('error:', response.data.message);
        }
    }).catch((error) => {
        console.log('error', error);
    })

    await sendRequest(path='/log', method='POST' , data = {command:'downloadOne', path:dir[0]}).then((response) => {
        if (response.statusCode === 200){
            console.log('statusCode:', response.statusCode);
            console.log('message:', JSON.parse(response.data).message);
        }else {
            console.log('statusCode:', response.statusCode);
            console.log('error:', response.data.message);
        }
    }).catch((error) => {
        console.log('error', error);
    })

    // await sendRequest(path='/log', method='POST' , data = {command:'downloadAll'}).then((response) => {
    //     if (response.statusCode === 200){
    //         console.log('statusCode:', response.statusCode);
    //         console.log('message:', JSON.parse(response.data).message);
    //     }else {
    //         console.log('statusCode:', response.statusCode);
    //         console.log('error:', response.data.message);
    //     }
    // }).catch((error) => {
    //     console.log('error', error);
    // });
    //
    // await sendRequest(path='/log', method='POST' , data = {command:'deleteOne', path:dir[0]}).then((response) => {
    //     if (response.statusCode === 200){
    //         console.log('statusCode:', response.statusCode);
    //         console.log('message:', JSON.parse(response.data).message);
    //     }else {
    //         console.log('statusCode:', response.statusCode);
    //         console.log('error:', response.data.message);
    //     }
    // }).catch((error) => {
    //     console.log('error', error);
    // });
    // await sendRequest(path='/log', method='POST' , data = {command:'deleteAll'}).then((response) => {
    //     if (response.statusCode === 200){
    //         console.log('statusCode:', response.statusCode);
    //         console.log('message:', JSON.parse(response.data).message);
    //     }else {
    //         console.log('statusCode:', response.statusCode);
    //         console.log('error:', response.data.message);
    //     }
    // }).catch((error) => {
    //     console.log('error', error);
    // });
}
main();


test/moduleTest/diskInfo.js
var diskinfo = require('diskinfo');
//当前盘符
var current_disk = __dirname.substr(0,2).toLowerCase();

//获得所有磁盘空间
diskinfo.getDrives(function(err, aDrives) {
    //遍历所有磁盘信息
    for (var i = 0; i < aDrives.length; i++) {
        //只获取当前磁盘信息
        if( aDrives[i].mounted.toLowerCase() == current_disk ){
            //盘符号
            var mounted = 'mounted ' + aDrives[i].mounted;
            //总量
            var total = 'total ' + (aDrives[i].blocks /1024 /1024 /1024).toFixed(0) + "gb";
            //已使用
            var used = 'used ' + (aDrives[i].used /1024 /1024 /1024).toFixed(0) + "gb";
            //可用
            var available = 'available ' + (aDrives[i].available /1024 /1024 /1024).toFixed(0) + "gb";
            //使用率
            var capacity = 'capacity ' + aDrives[i].capacity;
            console.log(mounted, total, used, available, capacity);
        }
    }
});

app.js
/**
 * 原创性检测服务器
 */

const path = require('path');
/**
 * web框架
 * **/
const koa = require('koa');
/**
 * HTTPS
 * **/
const https = require('https');                 // https
const sslify = require('koa-sslify').default;   // 强制使用HTTPS
const static = require('koa-static');           // 静态服务器
const koaBody = require('koa-body');
const bodyParser = require('koa-bodyparser');
/**
 * 文件模块
 * **/
const logHelper = require('./controllers/logHelper.js') // 文件模块 日志
const schedule = require('node-schedule');              // 文件模块 定时器
const os = require('os');                               // 文件模块 系统
const diskinfo = require('diskinfo');                   // 文件模块 磁盘信息
const nodemailer = require('nodemailer');               // 文件模块 邮箱
/**
 * 路由
 * **/
const imageDetection = require('./controllers/imageDetection.js')
const log = require('./controllers/log.js')
const monitor = require('./controllers/monitor.js')
/**
 * 配置文件
 * **/
const config = require('./config.json');

const app = new koa();

app.use(koaBody({
    multipart: true,
    formLimit:"2mb",
    jsonLimit:"2mb"
}));
app.use(bodyParser({
    formLimit:"2mb",
    jsonLimit:"2mb"
}));
app.use(sslify());                                       // 强制使用HTTPS
// app.use(static(path.join(__dirname, 'logs')));           // 静态服务器
// app.use(static(path.join(__dirname, 'assets/images')));  // 静态服务器
// app.use((ctx)=>{
//     console.log("hello");
// });
/**
 * 路由
 * **/
app.use(imageDetection.routes())
    .use(log.routes())
    .use(monitor.routes())

let options = {
    key: config.privateKey,
    cert: config.publicKey,
    ca: config.caCertificateList,
    // 要求验证客户端的证书
    requestCert: true,
    // 如果不是可信CA颁发的证书则断开连接
    rejectUnauthorized: true
};

https.createServer(options, app.callback()).listen(config.port, (error) => {
    if (error) {
        processError("https server launch error:", error.toString());
    } else {
        processInfo('imageDetection server running on:', 'https://' + config.host + ':' + config.port+'/imageDetection');
        processInfo('logs server running on:', 'https://' + config.host + ':' + config.port+'/logs');
        processInfo('monitor server running on:', 'https://' + config.host + ':' + config.port+'/monitor');
    }
});

/**
 * 进程事件
 * **/
process.on('warning', (warning) => {
    processWarn(warning.name);
    processWarn(warning.message);
    processWarn(warning.stack);
});

process.on('uncaughtException', (err, origin)=>{
    processError(err);
    processError(origin);
});

process.on('uncaughtException', (err, origin)=>{
    processError(err);
    processError(origin);
});

process.on('uncaughtExceptionMonitor', (err, origin)=>{
    processError(err);
    processError(origin);
});

process.on('unhandledRejection', (err, origin)=>{
    processError(err);
    processError(origin);
});

process.on('exit', (code) => {
    processInfo(`About to exit with code: ${code}`);
});

process.on('SIGINT', () => {
    processInfo('Received SIGINT. Press Control-D to exit.');
});

/**
 * 定时任务
 * **/
schedule.scheduleJob(config.monitor.schedule, async ()=>{
    // 1. 进程监控
    let {rss, heapTotal, heapUsed, external} = process.memoryUsage();
    rss = parseFloat((rss/1024/1024).toFixed(2));
    external = parseFloat((external/1024/1024).toFixed(2));
    let heapUsage = parseFloat((heapUsed/heapTotal).toFixed(2));
    // 2. 内存监控
    let memUsage = parseFloat((os.freemem()/os.totalmem()).toFixed(2));
    // 3. CPU占用监控
    let cpuUsage = await cpuAverage();
    // 4. 磁盘监控
    let mounted, diskTotal, diskUsed, diskAvailable, diskUsage;
    await getDiskUsage().then((data)=>{
        mounted = data.mounted;
        diskTotal = data.diskTotal;
        diskUsed = data.diskUsed;
        diskAvailable = data.diskAvailable;
        diskUsage = data.diskUsage;
    },(error)=>{
        processError('schedule', error);
    });
    diskTotal = parseInt(diskTotal);
    diskUsed = parseInt(diskUsed);
    diskAvailable = parseInt(diskAvailable);
    diskUsage = parseFloat(diskUsage)/100;

    monitorDebug('进程监控', rss, external, heapTotal, heapUsage);
    monitorDebug('内存监控', os.totalmem(), memUsage);
    monitorDebug('CPU占用监控', cpuUsage);
    monitorDebug('磁盘监控', mounted, diskTotal, diskUsed, diskAvailable, diskUsage);

    let html;
    if (heapUsage > config.monitor.heapUsage[0]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>heapUsage>"+config.monitor.heapUsage[0]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', '进程异常', heapUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', '进程异常', heapUsage, error);
        });
    }
    if (heapUsage > config.monitor.heapUsage[1]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>heapUsage>"+config.monitor.heapUsage[1]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', '进程异常', heapUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', '进程异常', heapUsage, error);
        });
    }
    if (memUsage > config.monitor.memUsage[0]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>memUsage>"+config.monitor.memUsage[0]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', '内存异常', memUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', '内存异常', memUsage, error);
        });
    }
    if (memUsage > config.monitor.memUsage[1]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>memUsage>"+config.monitor.memUsage[1]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', '内存异常', memUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', '内存异常', memUsage, error);
        });
    }
    if (cpuUsage > config.monitor.cpuUsage[0]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>cpuUsage>"+config.monitor.cpuUsage[0]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', 'CPU占用异常', cpuUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', 'CPU占用异常', cpuUsage, error);
        });
    }
    if (cpuUsage > config.monitor.cpuUsage[1]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>cpuUsage>"+config.monitor.cpuUsage[1]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', 'CPU占用异常', cpuUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', 'CPU占用异常', cpuUsage, error);
        });
    }
    if (diskUsage > config.monitor.diskUsage[0]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>diskUsage>"+config.monitor.diskUsage[0]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', '磁盘异常', cpuUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', '磁盘异常', cpuUsage, error);
        });
    }
    if (diskUsage > config.monitor.diskUsage[1]){
        html = "<p>"+config.host+"异常</p>";
        html += "<p>diskUsage>"+config.monitor.diskUsage[1]+"</p>";
        sendEMail(config.monitor.emailAddress, config.host+"异常",html).then((data)=>{
            monitorWarn('sendEMail', '磁盘异常', cpuUsage, data);
        }).catch((error)=>{
            monitorError('sendEMail', '磁盘异常', cpuUsage, error);
        });
    }
    rss = undefined;
    heapTotal = undefined;
    heapUsed = undefined;
    external = undefined;
    heapUsage = undefined;
    memUsage = undefined;
    cpuUsage = undefined;
    mounted = undefined;
    diskTotal = undefined;
    diskUsed = undefined;
    diskAvailable = undefined;
    diskUsage = undefined;
    html = undefined;
});
/**
 * 函数
 * **/
const current_disk = __dirname.substr(0,2).toLowerCase();
const getDiskUsage = ()=>{
    return new Promise(((resolve, reject) => {
        try {
            diskinfo.getDrives(function(error, aDrives) {
                if (error){
                    reject(error);
                }
                for (let i = 0; i < aDrives.length; i++) {
                    if( aDrives[i].mounted.toLowerCase() == current_disk ){
                        let mounted = aDrives[i].mounted;
                        let diskTotal  = (aDrives[i].blocks /1024 /1024).toFixed(0);
                        let diskUsed = (aDrives[i].used /1024 /1024).toFixed(0);
                        let diskAvailable = (aDrives[i].available /1024 /1024).toFixed(0);
                        let diskUsage = aDrives[i].capacity;
                        resolve({mounted:mounted, diskTotal:diskTotal, diskUsed:diskUsed, diskAvailable:diskAvailable, diskUsage:diskUsage});
                        mounted = undefined;
                        diskTotal = undefined;
                        diskUsed = undefined;
                        diskAvailable = undefined;
                        diskTotal = undefined;
                        diskUsage = undefined;
                        break;
                    }
                }
            });
        }catch (error) {
            reject(error);
        }
    }));
}

const cpuAverage = ()=>{
    return new Promise(((resolve, reject) => {
        let cpu, cpus, idle, len, total, totalIdle, totalTick, type, cpuUsage=0;

        totalIdle = 0;
        totalTick = 0;
        cpus = os.cpus();
        len = cpus.length;
        for(let i=0;i<len;i++){
            cpu = cpus[i];
            for (type in cpu.times) {
                totalTick += cpu.times[type];
            }
            totalIdle += cpu.times.idle;
            idle = totalIdle / len;
            total = totalTick / len;
            cpuUsage += parseFloat(((total-idle)/total).toFixed(2));
        }
        cpuUsage += parseFloat((cpuUsage/len).toFixed(2));
        resolve(cpuUsage);
        cpu = undefined;
        cpus = undefined;
        idle = undefined;
        len = undefined;
        total = undefined;
        totalIdle = undefined;
        totalTick = undefined;
        totalTick = undefined;
        cpuUsage = undefined;
    }));
}
const sendEMail = (email, subject, html) => {
    return new Promise(((resolve, reject) => {
        let transporter = nodemailer.createTransport({
            service: 'qq',
            port: 465,
            secureConnection: true,
            auth: {
                user: '1191567189@qq.com',
                pass: 'ymhtyusyknzxiefe',
            }
        });

        let mailOptions = {
            from: '"智慧生活" <1191567189@qq.com>',
            // 843080741@qq.com
            to: email,
            subject: subject,
            html: html
        };
        transporter.sendMail(mailOptions, (error, info) => {
            if (error){
                reject(error);
            }
            resolve(info);
            transporter = undefined;
            mailOptions = undefined;
        });
    }))
}

/**
 * 日志 -- process
 * **/
const processDebug = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().debug('process', args);
    logHelper.getLogger('process_debug').debug(args);
}
const processInfo = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().info('process', args);
    logHelper.getLogger('process_info').info(args);
}
const processWarn = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().warn('process', args);
    logHelper.getLogger('process_warn').warn(args);
}
const processError = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().error('process', args);
    logHelper.getLogger('process_error').error(args);
}
/**
 * 日志 -- monitor
 * **/
const monitorDebug = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().debug('monitor', args);
    logHelper.getLogger('monitor_debug').debug(args);
}
const monitorInfo = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().info('monitor', args);
    logHelper.getLogger('monitor_info').info(args);
}
const monitorWarn = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().warn('monitor', args);
    logHelper.getLogger('monitor_warn').warn(args);
}
const monitorError = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().error('monitor', args);
    logHelper.getLogger('monitor_error').error(args);
}

controllers/monitor.js
const Router = require('koa-router');
/**
 * 核心模块
 * **/

/**
 * 文件模块
 * **/
const validator = require('validator');                 // 文件模块 数据验证模块
const logHelper = require('./logHelper.js');            // 文件模块 日志
const os = require('os');                               // 文件模块 系统
const diskinfo = require('diskinfo');                   // 文件模块 磁盘信息

const config = require('../config.json');   // 配置文件
/**
 * 路由
 * **/
const router = new Router();
router.post('/monitor',async (ctx,next)=>{
    // 1. 获取数据
    let userName = ctx.req.connection.getPeerCertificate().subject.CN;
    let url = ctx.request.url;
    let {command} = ctx.request.body;
    monitorInfo(url, userName, {command});
    // 2. 身份验证
    if (!config.monitor.userNames.includes(userName)) {
        returnJson(ctx, userName, 400, {message:'do not have permission.'});
        userName = undefined;
        url = undefined;
        command = undefined;
        return;
    }
    // 3. 参数验证
    if (validator.isEmpty(command)) {
        returnJson(ctx, userName, 400, {message:'command is null.'});
        userName = undefined;
        url = undefined;
        command = undefined;
        return;
    }
    let commands = ['getProcessState', 'getMemUsage', 'getCpuUsage', 'getDiskUsage', 'getAll'];
    if (!commands.includes(command)){
        returnJson(ctx, userName, 400, {message:'command error.'});
        userName = undefined;
        url = undefined;
        command = undefined;
        commands = undefined;
        return;
    }
    // 4.1 进程监控
    let {rss, heapTotal, heapUsed, external} = process.memoryUsage();
    rss = parseFloat((rss/1024/1024).toFixed(2));
    external = parseFloat((external/1024/1024).toFixed(2));
    let heapUsage = parseFloat((heapUsed/heapTotal).toFixed(2));
    monitorInfo(url, userName, '进程监控', rss, external, heapTotal, heapUsage);
    // 4.2 内存监控
    let totalmem = os.totalmem();
    let memUsage = parseFloat((os.freemem()/os.totalmem()).toFixed(2));
    monitorInfo(url, userName, '内存监控', totalmem, memUsage);
    // 4.3 CPU占用监控
    let cpuUsage = await cpuAverage();
    monitorInfo(url, userName, 'CPU占用监控', cpuUsage);
    // 4.4 磁盘监控
    let mounted, diskTotal, diskUsed, diskAvailable, diskUsage;
    await getDiskUsage().then((data)=>{
        mounted = data.mounted;
        diskTotal = data.diskTotal;
        diskUsed = data.diskUsed;
        diskAvailable = data.diskAvailable;
        diskUsage = data.diskUsage;
    },(error)=>{
        monitorError('getDiskUsage', error);
    });
    diskTotal = parseInt(diskTotal);
    diskUsed = parseInt(diskUsed);
    diskAvailable = parseInt(diskAvailable);
    diskUsage = parseFloat(diskUsage)/100;
    monitorInfo(url, userName, '磁盘监控', mounted, diskTotal, diskUsed, diskAvailable, diskUsage);

    // 5 处理命令
    if (command==='getProcessState'){
        returnJson(ctx, userName, 200, {rss:rss, external:external, heapTotal:heapTotal, heapUsage:heapUsage});
    }
    if (command==='getMemUsage'){
        returnJson(ctx, userName, 200, {totalmem:totalmem, memUsage:memUsage});
    }
    if (command==='getCpuUsage'){
        returnJson(ctx, userName, 200, {cpuUsage:cpuUsage});
    }
    if (command==='getDiskUsage'){
        returnJson(ctx, userName, 200, {diskTotal:diskTotal, diskUsage:diskUsage});
    }
    if (command==='getAll'){
        returnJson(ctx, userName, 200, {rss:rss, external:external, heapTotal:heapTotal, heapUsage:heapUsage,totalmem:totalmem, memUsage:memUsage, cpuUsage:cpuUsage, diskTotal:diskTotal, diskUsage:diskUsage});
    }
    userName = undefined;
    url = undefined;
    command = undefined;
    rss = undefined;
    heapTotal = undefined;
    heapUsed = undefined;
    external = undefined;
    heapUsage = undefined;
    totalmem = undefined;
    memUsage = undefined;
    cpuUsage = undefined;
    mounted = undefined;
    diskTotal = undefined;
    diskUsed = undefined;
    diskAvailable = undefined;
    diskUsage = undefined;
})

/**
 * 自定义函数
 * **/
const returnJson = (ctx, userName, code = 200, data = {}) => {
    /**
     * @description 回复客户端
     * @code   : 状态码
     * @data   : 消息(body)
     * @date   : 16:40 2021/9/25
     * @author : cups
     **/
    ctx.response.status = code;
    ctx.response.body = data;
    switch (code) {
        case 200:
            monitorInfo('returnJson', userName, data);
            break;
        case 400:
            monitorWarn('returnJson', userName, data);
            break;
        case 500:
            monitorError('returnJson', userName, data);
            break;
        default:
            monitorInfo('returnJson', userName, data);
            break;
    }
}

/**
 * 日志 -- monitor
 * **/
const monitorDebug = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().debug('monitor', args);
    logHelper.getLogger('monitor_debug').debug(args);
}
const monitorInfo = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().info('monitor', args);
    logHelper.getLogger('monitor_info').info(args);
}
const monitorWarn = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().warn('monitor', args);
    logHelper.getLogger('monitor_warn').warn(args);
}
const monitorError = (...args)=>{
    args = JSON.stringify(args);
    logHelper.getLogger().error('monitor', args);
    logHelper.getLogger('monitor_error').error(args);
}

/**
 * 函数
 * **/
const current_disk = __dirname.substr(0,2).toLowerCase();
const getDiskUsage = ()=>{
    return new Promise(((resolve, reject) => {
        try {
            diskinfo.getDrives(function(error, aDrives) {
                if (error){
                    reject(error);
                }
                for (let i = 0; i < aDrives.length; i++) {
                    if( aDrives[i].mounted.toLowerCase() == current_disk ){
                        let mounted = aDrives[i].mounted;
                        let diskTotal  = (aDrives[i].blocks /1024 /1024).toFixed(0);
                        let diskUsed = (aDrives[i].used /1024 /1024).toFixed(0);
                        let diskAvailable = (aDrives[i].available /1024 /1024).toFixed(0);
                        let diskUsage = aDrives[i].capacity;
                        resolve({mounted:mounted, diskTotal:diskTotal, diskUsed:diskUsed, diskAvailable:diskAvailable, diskUsage:diskUsage});
                        mounted = undefined;
                        diskTotal = undefined;
                        diskUsed = undefined;
                        diskAvailable = undefined;
                        diskTotal = undefined;
                        diskUsage = undefined;
                        break;
                    }
                }
            });
        }catch (error) {
            reject(error);
        }
    }));
}

const cpuAverage = ()=>{
    return new Promise(((resolve, reject) => {
        let cpu, cpus, idle, len, total, totalIdle, totalTick, type, cpuUsage=0;

        totalIdle = 0;
        totalTick = 0;
        cpus = os.cpus();
        len = cpus.length;
        for(let i=0;i<len;i++){
            cpu = cpus[i];
            for (type in cpu.times) {
                totalTick += cpu.times[type];
            }
            totalIdle += cpu.times.idle;
            idle = totalIdle / len;
            total = totalTick / len;
            cpuUsage += parseFloat(((total-idle)/total).toFixed(2));
        }
        cpuUsage += parseFloat((cpuUsage/len).toFixed(2));
        resolve(cpuUsage);
        cpu = undefined;
        cpus = undefined;
        idle = undefined;
        len = undefined;
        total = undefined;
        totalIdle = undefined;
        totalTick = undefined;
        totalTick = undefined;
        cpuUsage = undefined;
    }));
}
module.exports = router;

controllers/log.js
const Router = require('koa-router');
/**
 * 核心模块
 * **/
const fs = require("fs");                   // 核心模块 fs
const join = require('path').join;          // 核心模块 path
/**
 * 文件模块
 * **/
const validator = require('validator');     // 文件模块 数据验证模块
const logHelper = require('./logHelper.js') // 文件模块 日志

const config = require('../config.json');   // 配置文件
/**
 * 路由
 * **/
const router = new Router();
router.post('/log', async (ctx, next) => {
    // 1. 获取数据
    let userName = ctx.req.connection.getPeerCertificate().subject.CN;
    let url = ctx.request.url;
    let {command, path} = ctx.request.body;
    logInfo(url, userName, {command, path});
    // 2. 身份验证
    if (!config.log.userNames.includes(userName)) {
        returnJson(ctx, userName, 400, {message:'do not have permission.'});
        userName = undefined;
        url = undefined;
        command = undefined;
        path = undefined;
        return;
    }
    // 3. 参数验证
    let commands = ['getAllFileDir', 'downloadOne', 'downloadAll', 'deleteOne', 'deleteAll'];
    if (validator.isEmpty(command)) {
        returnJson(ctx, userName, 400, {message:'command is null.'});
        userName = undefined;
        url = undefined;
        command = undefined;
        path = undefined;
        commands = undefined;
        return;
    }
    if (!commands.includes(command)) {
        returnJson(ctx, userName, 400, {message:'command is not in ' + commands.toString() + '.'});
        userName = undefined;
        url = undefined;
        command = undefined;
        path = undefined;
        commands = undefined;
        return;
    }
    let paths = await readDirs(config.log.path);
    if (path!==undefined){
        if (!paths.includes(path)){
            returnJson(ctx, userName, 400, {message:'path error.'});
            userName = undefined;
            url = undefined;
            command = undefined;
            path = undefined;
            commands = undefined;
            return;
        }
    }
    // 4. 处理命令
    // 4.1 getAllFileDir
    if (command==='getAllFileDir'){
        returnJson(ctx, userName, 200, {message:paths});
    }
    // 4.2 downloadOne
    if (command==='downloadOne'){
        returnJson(ctx, userName, 200, {message:fs.readFileSync(path, 'utf8')});
    }
    // 4.3 downloadAll
    if (command==='downloadAll'){
        let data = [];
        for (let i=0;i<paths.length; i++){
            data.push(fs.readFileSync(paths[i], 'utf8'));
        }
        returnJson(ctx, userName, 200, {message:data});
    }
    // 4.4 deleteOne
    if (command==='deleteOne'){
        try {
            fs.unlinkSync(path);
        }catch (error) {
            returnJson(ctx, userName, 500, {message:error});
        }
        returnJson(ctx, userName, 200, {message:'OK'});
    }
    // 4.5 deleteAll
    if (command==='deleteAll'){
        try {
            for (let i=0;i<paths.length; i++){
                fs.unlinkSync(paths[i]);
            }
        }catch (error) {
            returnJson(ctx, userName, 500, {message:error});
        }
        returnJson(ctx, userName, 200, {message:'OK'});
    }
    userName = undefined;
    url = undefined;
    command = undefined;
    path = undefined;
    commands = undefined;
});

/**
 * 自定义函数
 * **/
const returnJson = (ctx, userName, code = 200, data = {}) => {
    /**
     * @description 回复客户端
     * @code   : 状态码
     * @data   : 消息(body)
     * @date   : 16:40 2021/9/25
     * @author : cups
     **/
    ctx.response.status = code;
    ctx.response.body = data;
    switch (code) {
        case 200:
            logInfo('returnJson', userName, data);
            break;
        case 400:
            logWarn('returnJson', userName, data);
            break;
        case 500:
            logError('returnJson', userName, data);
            break;
        default:
            logInfo('returnJson', userName, data);
            break;
    }
}

const logDebug = (...args) => {
    args = JSON.stringify(args);
    logHelper.getLogger().debug('log', args);
    logHelper.getLogger('log_debug').debug(args);
}
const logInfo = (...args) => {
    args = JSON.stringify(args);
    logHelper.getLogger().info('log', args);
    logHelper.getLogger('log_info').info(args);
}
const logWarn = (...args) => {
    args = JSON.stringify(args);
    logHelper.getLogger().warn('log', args);
    logHelper.getLogger('log_warn').warn(args);
}
const logError = (...args) => {
    args = JSON.stringify(args);
    logHelper.getLogger().error('log', args);
    logHelper.getLogger('log_error').error(args);
}

const readDirs = (path) => {
    /**
     * @description 读取文件列表
     * @param  : path
     * @return : dirs
     * @date   : 22:34 2021/9/26
     * @author : cups
     **/
    return new Promise((async (resolve) => {
        let jsonFiles = [];
        const findJsonFile = (path)=>{
            let files = fs.readdirSync(path);
            files.forEach(function (item, index) {
                let fPath = join(path,item);
                let stat = fs.statSync(fPath);
                if(stat.isDirectory() === true) {
                    findJsonFile(fPath);
                }
                if (stat.isFile() === true) {
                    jsonFiles.push(fPath);
                }
            });
        }
        findJsonFile(path);
        for (let i=0; i<jsonFiles.length; i++){
            jsonFiles[i] = './'+jsonFiles[i].replace('\\','/').replace('\\','/')
        }
        resolve(jsonFiles);
        jsonFiles = undefined;
    }));
}
module.exports = router;